File: app/src/main/java/com/davismiyashiro/expenses/view/splitter/SplitterPresenterImpl.java
package com.davismiyashiro.expenses.view.splitter;

import com.davismiyashiro.expenses.datatypes.Expense;
import com.davismiyashiro.expenses.datatypes.Participant;
import com.davismiyashiro.expenses.datatypes.Split;
import com.davismiyashiro.expenses.model.TabRepository;

import java.util.ArrayList;
import java.util.List;

public class SplitterPresenterImpl implements SplitterView.UserActionsListener {
    private SplitterView.View mView;
    private TabRepository mModel;

    private List<Participant> mParticipants;
    private List<Split> mSplits;

    public SplitterPresenterImpl (SplitterView.View splitterView, TabRepository model) {
        mView = splitterView;
        mModel = model;
    }

    @Override
    public void validateParticipantsSelected(int size) {
        boolean error = false;
        if (size == 0) {
            error = true;
            mView.setParticipantSelectedError();
        }
        if (!error)
            mView.splitWithGroup();
    }

    @Override
    public List <Participant> getParticipants (Expense expense) {

        if (expense!= null) {
            mModel.refreshData(); // Fix to get Participants list updated when switching tabs
            mModel.getParticipants(expense.getTabId(), new TabRepository.LoadParticipantsCallback() {
                @Override
                public void onParticipantsLoaded(List<Participant> participants) {
                    mParticipants = participants;
                }
            });
        }

        if (mParticipants == null) {
            mParticipants = new ArrayList<>();
        }
        return mParticipants;
    }

    public void saveSplits (List<Participant> payers, Expense expense) {

        mSplits = new ArrayList<>();

        //Delete Splits before adding new ones
        mModel.deleteSplitsByExpense(expense);

        for (int count = 0; count < payers.size(); count++){
            Split split = new Split (payers.get(count).getId(), expense.getId(), payers.size(), expense.getValue()/payers.size(), expense.getTabId());
            mSplits.add(split);
        }

        mModel.saveSplits(mSplits);
    }

}

File: app/src/main/java/com/davismiyashiro/expenses/view/splitter/SplitterPresenterImpl.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.view.splitter

import com.davismiyashiro.expenses.datatypes.Expense
import com.davismiyashiro.expenses.datatypes.Participant
import com.davismiyashiro.expenses.datatypes.Split
import com.davismiyashiro.expenses.model.Repository

class SplitterPresenterImpl(private val mView: SplitterView.View, private val mModel: Repository) : SplitterView.UserActionsListener {

    private var mParticipants = listOf<Participant>()
    private var mSplits = mutableListOf<Split>()

    override fun validateParticipantsSelected(size: Int) {
        var error = false
        if (size == 0) {
            error = true
            mView.setParticipantSelectedError()
        }
        if (!error)
            mView.splitWithGroup()
    }

    override fun getParticipants(expense: Expense): List<Participant> {

        if (expense != null) {
            mModel.refreshData() // Fix to get Participants list updated when switching tabs
            mModel.getParticipants(expense.tabId, object : Repository.LoadParticipantsCallback {
                override fun onParticipantsLoaded(participants: MutableList<Participant>) {
                    mParticipants = participants
                }
            })
        }

        return mParticipants
    }

    override fun saveSplits(payers: List<Participant>, expense: Expense) {

        // Delete Splits before adding new ones
        mModel.deleteSplitsByExpense(expense)

        for (count in payers.indices) {
            val split = Split(payers[count].id, expense.id, payers.size, expense.value / payers.size, expense.tabId)
            mSplits.add(split)
        }

        mModel.saveSplits(mSplits)
    }
}
